new setting proxy.skip_target_path_escaping in api def added

dencoded opened the pull request:
changes to fix https://github.com/TykTechnologies/tyk/issues/1374

as part of reverse proxying we [do escaping when we write actual request to upstream](https://golang.org/src/net/http/request.go?#L520) - we always do `req.URL.RequestURI()` which is doing escaping for path in request for upstream.

We don't have too much choice here. I've noticed that [RequestURI](https://golang.org/src/net/url/url.go?#L983) doesn't do any escaping if URL.Opaque has value, so I build a logic around it and introduced new `skip_target_path_escaping` flag which can be set in `proxy` section of API definition. It is a little bit hacky but I think it will work as URL struct is used here just in context of preparing URI for http request to be sent.

If we go with that approach we will need to document this param with note that customers should use it at their own risk (in case of some specific upstreams like SFCC) as all this RFCs about escaping were done on purpose.

buger commented:
@dencoded I added few new tests including failing one. I can't get it work like your tests, with `spec.Proxy.SkipTargetPathEscaping = false`, not sure why. 

Also, I changed Path to RawPath, because otherwise encoded urls get decoded.

dencoded commented on reverse_proxy.go:
```diff
@@ -253,6 +250,10 @@ func TykNewSingleHostReverseProxy(target *url.URL, spec *APISpec) *ReverseProxy
 		}
+
+		if spec.Proxy.SkipTargetPathEscaping {
+			req.URL.Opaque = req.URL.RawPath // if we set opaque here it will force URL.RequestURI to skip escaping
```
I think this change is not gonna work as `req.URL.RawPath` contains at this point full initial request URI - listen path + the rest of the path

dencoded replied:
another note - `req.URL.RawPath` is going to have query string (if any) `?abc=xyz` and `RequestURI()` is adding query string here to result of escape-call again: https://golang.org/src/net/url/url.go#L994
if I understand right - we will have get passed to upstream result of `req.URL.Opaque + "?" + u.RawQuery` with `?abc=xyz` doubled.

buger replied:
From what I understand RawPath do not include query string

dencoded replied:
ok, lets scratch my comment about query string - it is not included to `RawPath`
I still see issue with setting `req.URL.Opaque = req.URL.RawPath` as it contains at this point initial request path (incoming to gateway) and leads to sending it to upstream as is.

the changes https://github.com/TykTechnologies/tyk/commit/a7a2a04107400cc7f1b60fce81b2ae945d11bd04 @lonelycode submitted should address this issue as we remove listen path, and replace in requests URL with object (result of parsing striped URI)

buger replied:
Pls update your PR to use martin changes if you thin they will work.

dencoded replied:
👍

lonelycode commented:
Did you take a look at my branch @dencoded?

dencoded commented:
@lonelycode yes, I am/was looking at it. still need to figure out the idea behind it

dencoded commented:
@buger I've fixed tests you added - there were a little typo in test data. also, I've added some logic to make use of `URL.RawPath`:
- strip listen path from `RawPath` when it is needed as we use its value to direct request to upstream
- some times `URL.RawPath` [can be empty](https://golang.org/src/net/url/url.go#L628) after parsing incoming requests' URL

plus added more tests for combinations of `strip_listen_path` and `skip_target_path_escaping `

so, things left:
- bring back test coverage (it is -0.5% now)
- add special case when one of Path segments has URL like `/abc/http://test.com`, we need somehow to analyze URL, split in segments with respect to possible URLs in them (split by `/` won't work) and apply to each segments [PathEscape](https://golang.org/src/net/url/url.go#L270)
